# Week 2 Notes - Getting and Cleaning Data

## Reading from MySQL
Data in SQL are structured into databases -> databases consist of tables with fields -> tables contain entries as rows. The tables themselves often represent specific aspects of the data which are interlinked within the database - say a table for the salaries of employees, another of the annnual leave, another table for their personal details and so on. 

## Let's install MySQL
In R `install.packages("RMySQL")` 

In julia

In [1]:
using Pkg; Pkg.add("MySQL") ; using MySQL

   Resolving package versions...
  No Changes to `~/.julia/environments/v1.10/Project.toml`
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


In [2]:
Pkg.add("DataFrames") ; using DataFrames  

   Resolving package versions...
  No Changes to `~/.julia/environments/v1.10/Project.toml`
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


## Connecting to databases - UCSC genome browser example

Connecting to the UCSC MySQL server and pulling the databases on UCSC that are available to us -- let's do it in R. This will establish a connection to the server - a stream, and then using this stream we will execute a MySQL command `show databases;` to retrieve the available databases, and then disconnect from the stream. 
```R
ucscDB <- dbConnect(MySQL(), user="genome", host="genome-mysql.cse.ucsc.edu")  
result <- dbGetQuery(ucscDB, "show databases;"); dbDisconnect(ucscDB);
```

In Julia we can do this by employing the MySQL.jl package - part of the Databases.jl family;

In [3]:
# Connecting https://mysql.juliadatabases.org/dev/
ucscDB = DBInterface.connect(MySQL.Connection, "genome-mysql.soe.ucsc.edu", "genome")

MySQL.Connection(host="genome-mysql.soe.ucsc.edu", user="genome", port="3306", db="")

In [4]:
# Query the server and store the query in a dataframe - or a csv etc. 
result = DBInterface.execute(ucscDB, "show databases") |> DataFrame; 

In [5]:
# Lets view the result - we can see that it lists all the genomes stored on UCSC
result

340×1 DataFrame
 Row │ Database 
     │ String   
─────┼──────────
   1 │ acaChl1
   2 │ ailMel1
   3 │ allMis1
   4 │ allSin1
   5 │ amaVit1
   6 │ anaPla1
   7 │ ancCey1
   8 │ angJap1
   9 │ anoCar1
  10 │ anoCar2
  11 │ anoGam1
  ⋮  │    ⋮
 331 │ visiGene
 332 │ wuhCor1
 333 │ xenLae2
 334 │ xenTro1
 335 │ xenTro10
 336 │ xenTro2
 337 │ xenTro3
 338 │ xenTro7
 339 │ xenTro9
 340 │ zonAlb1
319 rows omitted

In [6]:
"hg38" in result.Database 

true

In [7]:
# Close the connection stream 
DBInterface.close!(ucscDB)

### Now that we've connected to the MySQL server, we will connect to a specific database and perform some queries. 

In R - we'll connect, retrieve all of the table associated with the db, and then execute a function to see how many table are stored 
```R
hg38 <- dbConnect(MySQL(), user="genome", db="hg38", host="genome-mysql.soe.ucsc.edu")
allTables <- dbListTables(hg38)
length(allTables)
```

Let's get cracking on Julia

In [8]:
hg38 = DBInterface.connect(MySQL.Connection, "genome-mysql.soe.ucsc.edu", "genome", db="hg38")

MySQL.Connection(host="genome-mysql.soe.ucsc.edu", user="genome", port="3306", db="hg38")

In [9]:
hg38Tables = DBInterface.execute(hg38, "show tables") |> DataFrame; 

In [10]:
hg38Tables[1:10, :]

10×1 DataFrame
 Row │ Tables_in_hg38      
     │ String              
─────┼─────────────────────
   1 │ affyGnf1h
   2 │ affyU133
   3 │ affyU95
   4 │ all_est
   5 │ all_mrna
   6 │ all_sts_primer
   7 │ all_sts_seq
   8 │ altLocations
   9 │ altSeqLiftOverPsl
  10 │ altSeqLiftOverPslP3

A whopping 2835 different tables ! UCSC is extreeeeemly information rich, as we can see, there are many many many data sources we can pull from.    

### Now to investigate the specific fields within a specific table

In R; get a table and perform a basic SQL function to show how many fields are in the table
```R
dbListFields(hg38, "all_mrna")
dbGetQuery(hg38, "select count(*) from all_mrna")
```

In Julia

In [11]:
countsql = DBInterface.execute(hg38, "select count(*) from all_mrna") |> DataFrame; 

In [12]:
countsql

1×1 DataFrame
 Row │ count(*) 
     │ Int64    
─────┼──────────
   1 │ 10489979

Now lets play with the contents of the fields in R

```R
mrnaData <- dbReadTable(hg38, "all_mrna")
head(mrnaData)
```

Julia !

In [30]:
mrnaData = DBInterface.execute(hg38, "select * from all_mrna;") |> DataFrame

10489979×22 DataFrame
      Row │ bin     matches  misMatches  repMatches  nCount  qNumInsert  qBase ⋯
          │ UInt16  UInt32   UInt32      UInt32      UInt32  UInt32      UInt3 ⋯
──────────┼─────────────────────────────────────────────────────────────────────
        1 │    585     1579          25           0       0           0        ⋯
        2 │    585     1419          21           0       0           0
        3 │    585     1533          12           0       0           0
        4 │    585     1578          27           0       0           0
        5 │    585     1652           0           0       0           0        ⋯
        6 │    585     1650           2           0       0           0
        7 │    585     1648           4           0       0           0
        8 │    585     1485           3           0       0           0
        9 │    585     1631           8           0       0           0        ⋯
       10 │    585     1736           4           0       0           0
       11 │    585      402           0           0       0           0
    ⋮     │   ⋮        ⋮         ⋮           ⋮         ⋮         ⋮             ⋱
 10489970 │    586      287           0           0       0           0
 10489971 │    586      285           0           0       0           0        ⋯
 10489972 │    586      279          10           0       0           0
 10489973 │    586      285           1           0       0           0
 10489974 │    586      284           2           0       0           0
 10489975 │    587      298           4           0       0           0        ⋯
 10489976 │    587      272           4           0       0           0
 10489977 │    588      280           7           0       0           0
 10489978 │    588      283           3           0       0           0
 10489979 │    703     2427           0           0       0           0        ⋯
                                            16 columns and 10489958 rows omitted

In [29]:
Pkg.add("CSV") ; using CSV

   Resolving package versions...
  No Changes to `~/.julia/environments/v1.10/Project.toml`
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


### Refined queries 
MySQL has a wide range of query options which permit the extration of virtually any aspect of the data, with conditionals, ranges, mismatches and so on. These statements must simply be crafted according to the query structure and provided to the julia functions in order to get what you're looking for.    

For instance we can extract entries from the table which have values in the 'mismatches' column between 1 and 3;   

In R
```R
query <- dbSendQuery(hg38, "select * from all_mrna where misMatches between 1 and 3") 
mrnas <- fetch(query)
```

In Julia

In [32]:
query = DBInterface.execute(hg38, "select * from all_mrna where misMatches between 1 and 3") |> DataFrame

403833×22 DataFrame
    Row │ bin     matches  misMatches  repMatches  nCount  qNumInsert  qBaseIn ⋯
        │ UInt16  UInt32   UInt32      UInt32      UInt32  UInt32      UInt32  ⋯
────────┼───────────────────────────────────────────────────────────────────────
      1 │    585     1650           2           0       0           0          ⋯
      2 │    585     1485           3           0       0           0
      3 │    585      925           2           0       0           1
      4 │    585      986           1           0       0           0
      5 │    585      970           3           0       0           0          ⋯
      6 │    585      953           2           0       0           0
      7 │    585      775           1           0       0           0
      8 │    585      974           2           0       0           0
      9 │    585     1123           1           0       0           0          ⋯
     10 │    585     1122           2           0       0           0
     11 │    586     1381           1           0       0           0
   ⋮    │   ⋮        ⋮         ⋮           ⋮         ⋮         ⋮            ⋮  ⋱
 403824 │    755      292           3           0       0           0
 403825 │     94      321           3           0       0           0          ⋯
 403826 │     94      319           3           0       0           0
 403827 │    585      294           2           0       0           0
 403828 │    585      292           3           0       0           0
 403829 │    585      294           3           0       0           0          ⋯
 403830 │    585      284           3           0       0           0
 403831 │    586      285           1           0       0           0
 403832 │    586      284           2           0       0           0
 403833 │    588      283           3           0       0           0          ⋯
                                              16 columns and 403812 rows omitted

In [34]:
first(query)

DataFrameRow
 Row │ bin     matches  misMatches  repMatches  nCount  qNumInsert  qBaseInser ⋯
     │ UInt16  UInt32   UInt32      UInt32      UInt32  UInt32      UInt32     ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │    585     1650           2           0       0           0             ⋯
                                                              16 columns omitted

In [ ]:
size(query)

### Close the connection!

In [37]:
DBInterface.close!(hg38)

## HDF5 - Reading and Handling this data type
HDF5 is a data format used for storing large datasets - FAST5 uses a HDF5 backbone, and we know how large FAST5 files are!!! It stands for **H**eirarchical **D**ata **F**ormat. 

As expected, HDF5 support for R comes in a library, this time it is downloaded from bioconductor 
```R
source("url")
biocLite("rhdf5")
library(rhdf5)
```

We can create an example file using the hdf5 functions `created = h5createFile("example.h5")`    

Julia too has a package for working with HDF5 formats, can you guess the original name? **HDF5.jl**. Remarkable indeed. https://juliaio.github.io/HDF5.jl/stable/ - A nice explaner of HDF5 from the julia software page 

*"HDF5 stands for Hierarchical Data Format v5 and is closely modeled on file systems. In HDF5, a "group" is analogous to a directory, a "dataset" is like a file. HDF5 also uses "attributes" to associate metadata with a particular group or dataset. HDF5 uses ASCII names for these different objects, and objects can be accessed by Unix-like pathnames, e.g., "/sample1/tempsensor/firsttrial" for a top-level group "sample1", a subgroup "tempsensor", and a dataset "firsttrial"."*  

(Datasets cannot have child datasets, but groups can have either.) 

Groups/subgroups/datasets NOT Groups/datasets/datasets/groups

In [39]:
Pkg.add("HDF5") ; using HDF5

   Resolving package versions...
   Installed MPIPreferences ─ v0.1.10
   Installed HDF5_jll ─────── v1.12.2+2
   Installed Requires ─────── v1.3.0
   Installed HDF5 ─────────── v0.17.1
    Updating `~/.julia/environments/v1.10/Project.toml`
  [f67ccb44] + HDF5 v0.17.1
    Updating `~/.julia/environments/v1.10/Manifest.toml`
  [f67ccb44] + HDF5 v0.17.1
  [3da0fdf6] + MPIPreferences v0.1.10
  [ae029012] + Requires v1.3.0
⌃ [0234f1f7] + HDF5_jll v1.12.2+2
        Info Packages marked with ⌃ have new versions available and may be upgradable.
Precompiling project...
  ✓ MPIPreferences
  ✓ Requires
  ✓ HDF5_jll
  ✓ HDF5
  4 dependencies successfully precompiled in 10 seconds. 60 already precompiled.


To write an example file in julia

In [41]:
example_hd = h5open("example.h5", "cw")

🗂️ HDF5.File: (read-write) example.h5

In [43]:
close(example_hd)

In [45]:
example_hd = h5open("example.h5", "r+")

🗂️ HDF5.File: (read-write) example.h5

Since HDF5 files are hierarchical and based upon a file-system like structure, we create groups

Once we have groups, we write to the groups, and perhaps subgroups, subsubgroups etc. - remember, HDF5 is akin to a filesystem

We'll create some matrix data which we'll write to a group called foo/A in R
```R
A = matrix(1:10, nr=5, nc=2)
h5write(A, "example.h5", "foo/A") #group is the third argument
B = array(seq(0.1,2.0, by=0.1), dim=c(5,2,2))
attr(B, "scale") <- "liter"
h5write(B, "example.h5", "foo/foobaa/B") 
h5ls("example.h5") # h5 ls view 
```


In Julia let's work with the HDF5 package 

Create a group called "foo" 

In [46]:
create_group(example_hd, "foo")

📂 HDF5.Group: /foo (file: example.h5)

Create some mock data

In [49]:
samp = Array(rand(2, 4))

2×4 Matrix{Float64}:
 0.453924  0.774484  0.896302  0.714723
 0.141766  0.911775  0.745648  0.932475

Write it to the group foo

In [51]:
# If a group doesn't already exist we can write to it by indexing
example_hd["newgroup"] = "yes"

"yes"

In [66]:
# To write to a pre-existing group 
# First initialize the group and load it into a variable
g = example_hd["foo"]
# Write to it by indexing directly
g["mydataset"] = samp 
# Write to it using the create_dataset() function
create_dataset(g, "simplestring", zeros(1, 2)) 

(HDF5.Dataset: /foo/simplestring (file: example.h5 xfer_mode: 0), HDF5.Datatype: H5T_IEEE_F64LE)

In [60]:
example_hd["foo/mydataset"]

🔢 HDF5.Dataset: /foo/mydataset (file: example.h5 xfer_mode: 0)

Read the contents of the dataset and its groups using the **read** function

In [61]:
read(example_hd["foo/mydataset"])

2×4 Matrix{Float64}:
 0.453924  0.774484  0.896302  0.714723
 0.141766  0.911775  0.745648  0.932475

In [67]:
read(example_hd["foo/simplestring"])

1×2 Matrix{Float64}:
 0.0  0.0

In [68]:
example_hd

🗂️ HDF5.File: (read-write) example.h5
├─ 📂 foo
│  ├─ 🔢 mydataset
│  └─ 🔢 simplestring
└─ 🔢 newgroup

In [71]:
close(example_hd)

For convience and consistency we can also use the **do** block conventions, which will take care of closing the stream for us 

In [73]:
h5open("example.h5", "r+") do stream
    group = create_group(stream, "dogroup")
    dataset = create_dataset(group, "thisdata", Float64, (10,10))
    write(dataset, rand(10,10))
end 

Basic notes;
Datasets can be create by 
```julia
g["mydataset"] = rand(3,5)
# or
write(g, "mydataset", rand(3,5))
``` 

### Reading specific parts of the data 
In R; 
```R
h5read("example.h5", "foo/A")
h5read("example.h5", "foo/new/dataset")
```

In Julia

In [89]:
openh5 = h5open("example.h5", "r+")

🗂️ HDF5.File: (read-write) example.h5
├─ 📂 dogroup
│  └─ 🔢 thisdata
├─ 📂 foo
│  ├─ 🔢 mydataset
│  └─ 🔢 simplestring
└─ 🔢 newgroup

In [91]:
read(openh5,"foo/mydataset")

2×4 Matrix{Float64}:
 0.453924  0.774484  0.896302  0.714723
 0.141766  0.911775  0.745648  0.932475

### Chunking and Compression 
The section on the Julia HDF5.jl package site does a nice job of explaining this ! https://juliaio.github.io/HDF5.jl/stable/

## Webscraping HTML 
Scraping information from the internet is a fun endevour, which based upon the virility of ones creativity, can lead to a large number of exciting avenues of adventure. As the web has been around for decades, almost every high level programming language has its own packages and libraries for working with web HTML and XML files. 
The most important part of web scraping is to avoid excessive request to websites as this will likely get your IP address blocked. We can safely assume that large companies such as amazon are interested in protecting themselves from massive data trawling operations, even though they are doing it on a massive scale. 

### Examples using google scholar

In R first; 

We'll store a url link, and then simply read the lines of the link, and then close the link 
```R
con = url("https://scholar.google.com/citations?user=kzht3-0AAAAJ&hl=en")
htmlCode = readLines(con)
close(con)
htmlCode
```

Now in Julia - we can use the HTTP.jl package - or Requests.jl

In [105]:
Pkg.add("HTTP") ; using HTTP

   Resolving package versions...
   Installed ExceptionUnwrapping ─ v0.1.10
   Installed SimpleBufferStream ── v1.1.0
   Installed ConcurrentUtilities ─ v2.3.0
   Installed BitFlags ──────────── v0.1.8
   Installed OpenSSL ───────────── v1.4.1
   Installed LoggingExtras ─────── v1.0.3
   Installed URIs ──────────────── v1.5.1
   Installed HTTP ──────────────── v1.10.1
    Updating `~/.julia/environments/v1.10/Project.toml`
  [cd3eb016] + HTTP v1.10.1
    Updating `~/.julia/environments/v1.10/Manifest.toml`
  [d1d4a3ce] + BitFlags v0.1.8
  [f0e56b4a] + ConcurrentUtilities v2.3.0
  [460bff9d] + ExceptionUnwrapping v0.1.10
  [cd3eb016] + HTTP v1.10.1
  [e6f89c97] + LoggingExtras v1.0.3
  [4d8831e6] + OpenSSL v1.4.1
  [777ac1f9] + SimpleBufferStream v1.1.0
  [5c2747f8] + URIs v1.5.1
Precompiling project...
  ✓ BitFlags
  ✓ SimpleBufferStream
  ✓ ExceptionUnwrapping
  ✓ LoggingExtras
  ✓ URIs
  ✓ ConcurrentUtilities
  ✓ OpenSSL
  ✓ HTTP
  8 dependencies successfully precompiled in 7 seconds

In [94]:
url = "https://scholar.google.com/citations?user=kzht3-0AAAAJ&hl=en"

"https://scholar.google.com/citations?user=kzht3-0AAAAJ&hl=en"

In [113]:
resp = HTTP.get(url)

HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Date: Wed, 31 Jan 2024 02:32:23 GMT
Expires: Fri, 01 Jan 1990 00:00:00 GMT
Cache-Control: no-cache, must-revalidate
Pragma: no-cache
Content-Type: text/html; charset=ISO-8859-1
X-Content-Type-Options: nosniff
Content-Encoding: gzip
Server: citations
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
Transfer-Encoding: chunked

<!doctype html><html><head><title>Michael Lynch - Google Scholar</title><meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><meta name="referrer" content="always"><meta name="viewport" content="width=device-width,initial-scale=1,minimum-scale=1,maximum-scale=2"><meta name="format-detection" content="telephone=no"><link rel="shortcut icon" href="/favicon.ico"><link rel="canonical" href="https://scholar.google.com/citations?user=kzht3-0AAAAJ&amp;hl=en"><meta name="description" content="Arizona State

#### A note on GET vs POST requests
**GET**
* parameters are in the URL
* used for fetching and *GETTING* documents
* maximum URL length 
* OK to cache
* won't change the server

**POST**
* parameters are in the body
* used for updating and *POSTING* data 
* not ok to cache
* can change the server

#### Parsing HTML file as a XML - conversion 
Depending on the structure of the HTML, we may be able to translate it as a XML and interpret it in the same structure - let's give it a go and see if it works. 

In R; 
```R
library(XML)
url <- "https://scholar.google.com/citations?user=kzht3-0AAAAJ&hl=en"
html <- htmlTreeParse(url, useInternalNodes=T)
```
Using xpath language queries

```R
xpathSApply(html, "//title", xmlValue)
```

Now let's look at the citation counts
```R
xpathSApply(html, "//td[@id='col-citedby']", xmlValue)
``` 

Give it a crack in Julia using the EzXML package - it will take a few more steps, and perhaps it can get compressed down with experience into some simpler code.

**The basic steps are**
1. Load HTTP package, perform a HTTP get request on the url
2. Parse the URL as a string by accessing the requests body 
3. Using the EzXML package, parse the html string using parsehtml()
4. Use the xpath query language with findall() and nodecontent() to extract the relevant information from the file


In [114]:
Pkg.add("EzXML") ; using EzXML

   Resolving package versions...
   Installed XML2_jll ─ v2.12.2+0
   Installed EzXML ──── v1.2.0
    Updating `~/.julia/environments/v1.10/Project.toml`
  [8f5d6c58] + EzXML v1.2.0
    Updating `~/.julia/environments/v1.10/Manifest.toml`
  [8f5d6c58] + EzXML v1.2.0
  [02c8fc9c] + XML2_jll v2.12.2+0
Precompiling project...
  ✓ XML2_jll
  ✓ EzXML
  2 dependencies successfully precompiled in 2 seconds. 72 already precompiled.


Open a HTTP request, and then parse the HTML file as a string which can be read by the EzXML package using the **parsehtml()** function. 

In [144]:
samp = HTTP.get(url, cookies=true);
data = String(samp.body)

"<!doctype html><html><head><title>Michael Lynch - Google Scholar</title><meta http-equiv=\"Content-Type\" content=\"text/html;charset=ISO-8859-1\"><meta http-equiv=\"X-UA-Compatible\" content=\"IE=Edge\"><meta name=\"referrer\" content=\"always\"><meta name=\"viewport\" content=\"wid" ⋯ 151840 bytes ⋯ "le=\"menuitem\" href=\"/intl/en/scholar/about.html\" tabindex=\"-1\" class=\"gs_md_li\">About Scholar</a><a role=\"menuitem\" href=\"//support.google.com/websearch?p=scholar_dsa&amp;hl=en&amp;oe=ASCII\" tabindex=\"-1\" class=\"gs_md_li\">Search help</a></div></div></div></body></html>"

Parse the html file now and store the root as a variable

In [154]:
q = parsehtml(data)
scholar_root = root(q)

┌ Warning: XMLError: Unexpected end tag : div from HTML parser (code: 76, line: 37)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97
┌ Warning: XMLError: Unexpected end tag : div from HTML parser (code: 76, line: 41)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97
┌ Warning: XMLError: Unexpected end tag : div from HTML parser (code: 76, line: 41)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97
┌ Warning: XMLError: htmlParseEntityRef: expecting ';' from HTML parser (code: 23, line: 75)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97
┌ Warning: XMLError: htmlParseEntityRef: expecting ';' from HTML parser (code: 23, line: 75)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97
┌ Warning: XMLError: htmlParseEntityRef: expecting ';' from HTML parser (code: 23, line: 75)
└ @ EzXML ~/.julia/packages/EzXML/DL8na/src/error.jl:97


EzXML.Node(<ELEMENT_NODE[html]@0x0000000005486f00>)

Using the **nodecontent.()** function of EzXML, find your query using the xpath language 

In [160]:
nodecontent.(findall("//title", scholar_root))

1-element Vector{String}:
 "Michael Lynch - Google Scholar"

Now for the more elaborate Xpath query

In [164]:
for citation in nodecontent.(findall("//td[@id='col-citedby']", scholar_root))
    println(citation)
end 

#### Using the httr library for R 
httr makes doing much of this a bit easier - as we can see the workflow is very similar to that of HTTP.jl 
```R
library(httr); html2 = GET(url)
content2 = content(html2, as="text")
parsedHtml = htmlParse(content2, asText=True)
xpathSapply(parsedHtml, "//title", xmlValue)
```

### Websites with user and password authentication
In order to access websites which request a user and password, in R we can include the information as an argument
```R
pp2 = GET("url", authenticate("user","passwd"))
```

In Julia we prefix the server address as such;

In [168]:
pp2 = HTTP.get("https://user:passwd@httpbin.org/basic-auth/user/passwd")

HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Date: Thu, 01 Feb 2024 02:37:55 GMT
Content-Type: application/json
Content-Length: 47
Connection: keep-alive
Server: gunicorn/19.9.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true

{
  "authenticated": true, 
  "user": "user"
}
"""

## Quiz 
1. Read up on the github API 
https://github.com/settings/applications
. Access the API to get information on your instructors repositories (hint: this is the url you want "https://api.github.com/users/jtleek/repos"). Use this data to find the time that the datasharing repo was created. What time was it created?   

This tutorial may be useful (
https://github.com/hadley/httr/blob/master/demo/oauth2-github.r
). You may also need to run the code in the base R package and not R studio.

In Julia, we'll use the github.jl package from https://github.com/JuliaWeb/GitHub.jl 

In [174]:
Pkg.add("GitHub") ; using GitHub

   Resolving package versions...
  No Changes to `~/.julia/environments/v1.10/Project.toml`
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


Using julias github package, let's take a look at all of jtleeks repos, using the simple **repos** function with his username as the argument 

In [197]:
leeks_repos = repos("jtleek")

(Repo[Repo (all fields are Union{Nothing, T}):
  name: "2018"
  full_name: "jtleek/2018"
  description: "Fall 2018 repository with course materials for JHU Advanced Data Science"
  language: "HTML"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 155565363
  size: 60855
  forks_count: 3
  stargazers_count: 1
  watchers_count: 1
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/2018")
  html_url: URI("https://github.com/jtleek/2018")
  clone_url: URI("https://github.com/jtleek/2018.git")
  ssh_url: URI("git@github.com:jtleek/2018.git")
  homepage: URI("https://jhu-advdatasci.github.io/2018/")
  pushed_at: DateTime("2018-10-30T18:13:41")
  created_at: DateTime("2018-10-31T13:50:37")
  updated_at: DateTime("2021-12-05T09:30:00")
  has_issues: false
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: true, Repo (all fields are Union{Nothing, T}):
  name: "ads2020"
  full_name: "jtleek/ads2020"
  description: "Advanced Data Scienc

In [210]:
# Doesn't work - I can't seem to get an iterator going
for f in leeks_repos
        print(f)
    end 
end 

Repo[Repo (all fields are Union{Nothing, T}):
  name: "2018"
  full_name: "jtleek/2018"
  description: "Fall 2018 repository with course materials for JHU Advanced Data Science"
  language: "HTML"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 155565363
  size: 60855
  forks_count: 3
  stargazers_count: 1
  watchers_count: 1
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/2018")
  html_url: URI("https://github.com/jtleek/2018")
  clone_url: URI("https://github.com/jtleek/2018.git")
  ssh_url: URI("git@github.com:jtleek/2018.git")
  homepage: URI("https://jhu-advdatasci.github.io/2018/")
  pushed_at: DateTime("2018-10-30T18:13:41")
  created_at: DateTime("2018-10-31T13:50:37")
  updated_at: DateTime("2021-12-05T09:30:00")
  has_issues: false
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: true, Repo (all fields are Union{Nothing, T}):
  name: "ads2020"
  full_name: "jtleek/ads2020"
  description: "Advanced Data Science

  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "capitalIn21stCenturyinR"
  full_name: "jtleek/capitalIn21stCenturyinR"
  description: "Piketty in R"
  language: "HTML"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 20234724
  size: 374812
  forks_count: 128
  stargazers_count: 213
  watchers_count: 213
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/capitalIn21stCenturyinR")
  html_url: URI("https://github.com/jtleek/capitalIn21stCenturyinR")
  clone_url: URI("https://github.com/jtleek/capitalIn21stCenturyinR.git")
  ssh_url: URI("git@github.com:jtleek/capitalIn21stCenturyinR.git")
  pushed_at: DateTime("2016-07-18T17:22:51")
  created_at: DateTime("2014-05-27T20:38:31")
  updated_at: DateTime("2024-01-04T22:12:27")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: true
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):

  url: URI("https://api.github.com/repos/jtleek/day1")
  html_url: URI("https://github.com/jtleek/day1")
  clone_url: URI("https://github.com/jtleek/day1.git")
  ssh_url: URI("git@github.com:jtleek/day1.git")
  pushed_at: DateTime("2017-07-11T15:25:20")
  created_at: DateTime("2017-07-10T21:44:28")
  updated_at: DateTime("2017-07-11T14:13:55")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "derfinder"
  full_name: "jtleek/derfinder"
  language: "R"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 11549405
  size: 388
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/derfinder")
  html_url: URI("https://github.com/jtleek/derfinder")
  clone_url: URI("https://github.com/jtleek/derfinder.git")
  ssh_url: URI("git@github.com:jtleek/derfinder.git")
  pushed_at: DateTime("2013-06-24T21:17:27")
 

  size: 41702
  forks_count: 11
  stargazers_count: 16
  watchers_count: 16
  open_issues_count: 1
  url: URI("https://api.github.com/repos/jtleek/genstats_site")
  html_url: URI("https://github.com/jtleek/genstats_site")
  clone_url: URI("https://github.com/jtleek/genstats_site.git")
  ssh_url: URI("git@github.com:jtleek/genstats_site.git")
  pushed_at: DateTime("2015-09-07T13:54:52")
  created_at: DateTime("2015-08-20T12:23:21")
  updated_at: DateTime("2023-10-05T18:12:28")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: true
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "github-slideshow"
  full_name: "jtleek/github-slideshow"
  description: "A robot powered training repository :robot:"
  language: "Ruby"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 289258980
  size: 3515
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 1
  url: URI("https://api.github.com/repos/jtleek/github-slides

  html_url: URI("https://github.com/jtleek/jhsph753and4")
  clone_url: URI("https://github.com/jtleek/jhsph753and4.git")
  ssh_url: URI("git@github.com:jtleek/jhsph753and4.git")
  pushed_at: DateTime("2014-05-15T14:22:41")
  created_at: DateTime("2014-01-04T21:06:44")
  updated_at: DateTime("2022-04-14T23:52:19")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: true
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "jhudash"
  full_name: "jtleek/jhudash"
  description: "A repository for all things DaSH"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 42834789
  size: 381
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/jhudash")
  html_url: URI("https://github.com/jtleek/jhudash")
  clone_url: URI("https://github.com/jtleek/jhudash.git")
  ssh_url: URI("git@github.com:jtleek/jhudash.git")
  pushed_at: DateTime("2015-09-21T10:35:40")
  created_a

  html_url: URI("https://github.com/jtleek/newproject")
  clone_url: URI("https://github.com/jtleek/newproject.git")
  ssh_url: URI("git@github.com:jtleek/newproject.git")
  pushed_at: DateTime("2023-08-08T04:37:22")
  created_at: DateTime("2017-08-30T18:14:23")
  updated_at: DateTime("2021-11-21T14:36:49")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "new_project"
  full_name: "jtleek/new_project"
  description: "This is my new project"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 101907019
  size: 0
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/new_project")
  html_url: URI("https://github.com/jtleek/new_project")
  clone_url: URI("https://github.com/jtleek/new_project.git")
  ssh_url: URI("git@github.com:jtleek/new_project.git")
  pushed_at: DateTime("2017-08-30T16:58:22")
  c

  full_name: "jtleek/rmd4edu"
  description: "A fresh batch of R Markdown templates"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 273780186
  size: 12873
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/rmd4edu")
  html_url: URI("https://github.com/jtleek/rmd4edu")
  clone_url: URI("https://github.com/jtleek/rmd4edu.git")
  ssh_url: URI("git@github.com:jtleek/rmd4edu.git")
  pushed_at: DateTime("2019-09-03T14:08:12")
  created_at: DateTime("2020-06-20T20:33:17")
  updated_at: DateTime("2023-09-08T18:09:02")
  has_issues: false
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: true, Repo (all fields are Union{Nothing, T}):
  name: "robotjeff"
  full_name: "jtleek/robotjeff"
  description: "This is the Shiny app to make robot Jeff talk"
  language: "R"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 94168246
  size: 2
  forks_count: 0
  stargazers_count

  has_wiki: true
  has_downloads: true
  has_pages: true
  license: License("NOASSERTION")
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "slipper"
  full_name: "jtleek/slipper"
  description: "Tidy and easy bootstrapping"
  language: "R"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 75975252
  size: 74
  forks_count: 12
  stargazers_count: 119
  watchers_count: 119
  open_issues_count: 2
  url: URI("https://api.github.com/repos/jtleek/slipper")
  html_url: URI("https://github.com/jtleek/slipper")
  clone_url: URI("https://github.com/jtleek/slipper.git")
  ssh_url: URI("git@github.com:jtleek/slipper.git")
  pushed_at: DateTime("2017-10-05T18:12:29")
  created_at: DateTime("2016-12-08T21:08:11")
  updated_at: DateTime("2024-01-24T22:26:21")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "software"
  full_name: "jtleek/software"
  de

  created_at: DateTime("2015-07-05T12:57:04")
  updated_at: DateTime("2015-07-05T12:57:04")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "testproject"
  full_name: "jtleek/testproject"
  default_branch: "master"
  owner: Owner("jtleek")
  id: 38634839
  size: 112
  forks_count: 0
  stargazers_count: 0
  watchers_count: 0
  open_issues_count: 0
  url: URI("https://api.github.com/repos/jtleek/testproject")
  html_url: URI("https://github.com/jtleek/testproject")
  clone_url: URI("https://github.com/jtleek/testproject.git")
  ssh_url: URI("git@github.com:jtleek/testproject.git")
  pushed_at: DateTime("2015-07-06T17:37:08")
  created_at: DateTime("2015-07-06T17:34:35")
  updated_at: DateTime("2015-07-06T17:34:35")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false, Repo (all fields are Union{Nothing, T}):
  name: "testrepo

LoadError: ParseError:
# Error @ ]8;;file:///home/number25/MEGA/Computational-Bio/Kobe-n-Pascal/ComputationalBiology-for-Autodidacts/Course_notes/Coursera_Data_Science_Track_JHOP/Mod3_Data_Cleaning/Week_2/In[210]#5:1\In[210]:5:1]8;;\
    end 
end 
└─┘ ── invalid identifier

Search the output for the repo "datasharing" - hmmm this seems more difficult than I first assumed, as the output type is a Repo-Union custom type it seems 

Let's try honing in a specific repo, so use the **repo** function with the exact repo as the argument (including his username: username/repo)

In [184]:
repo("jtleek/datasharing")

Repo (all fields are Union{Nothing, T}):
  name: "datasharing"
  full_name: "jtleek/datasharing"
  description: "The Leek group guide to data sharing "
  default_branch: "master"
  owner: Owner("jtleek")
  id: 14204342
  size: 590
  subscribers_count: 561
  forks_count: 243587
  stargazers_count: 6434
  watchers_count: 6434
  open_issues_count: 892
  url: URI("https://api.github.com/repos/jtleek/datasharing")
  html_url: URI("https://github.com/jtleek/datasharing")
  clone_url: URI("https://github.com/jtleek/datasharing.git")
  ssh_url: URI("git@github.com:jtleek/datasharing.git")
  pushed_at: DateTime("2024-01-05T04:49:32")
  created_at: DateTime("2013-11-07T13:25:07")
  updated_at: DateTime("2024-02-01T12:03:04")
  has_issues: true
  has_wiki: true
  has_downloads: true
  has_pages: false
  private: false
  fork: false

Answer to the question is **created_at: DateTime("2013-11-07T13:25:07")**
 

## 2. 
The sqldf package allows for execution of SQL commands on R data frames. We will use the sqldf package to practice the queries we might send with the dbSendQuery command in RMySQL. 

Download the American Community Survey data and load it into an R object called  

https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2Fss06pid.csv

Which of the following commands will select only the data for the probability weights pwgtp1 with ages less than 50?

In [211]:
using CSV

Download the csv and port it into a dataframe 

In [213]:
download("https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2Fss06pid.csv"
, "quiz2.csv")

"quiz2.csv"

In [215]:
quiz2_dataframe = CSV.File("quiz2.csv") |> DataFrame

14931×239 DataFrame
   Row │ RT       SERIALNO  SPORDER  PUMA   ST     ADJUST   PWGTP  AGEP   CIT  ⋯
       │ String1  Int64     Int64    Int64  Int64  Int64    Int64  Int64  Int6 ⋯
───────┼────────────────────────────────────────────────────────────────────────
     1 │ P             186        1    700     16  1015675     89     43       ⋯
     2 │ P             186        2    700     16  1015675     92     42
     3 │ P             186        3    700     16  1015675    107     16
     4 │ P             186        4    700     16  1015675     91     14
     5 │ P             306        1    700     16  1015675    309     29       ⋯
     6 │ P             395        1    100     16  1015675    108     40
     7 │ P             395        2    100     16  1015675     90     15
     8 │ P             506        1    700     16  1015675    239     28
     9 │ P             506        2    700     16  1015675    213     30       ⋯
    10 │ P             506        3    700     16  1015675    219      4
    11 │ P             506        4    700     16  1015675    226      1
   ⋮   │    ⋮        ⋮         ⋮       ⋮      ⋮       ⋮       ⋮      ⋮      ⋮  ⋱
 14922 │ P         1357238        1    300     16  1015675    111     24
 14923 │ P         1357238        2    300     16  1015675     74     21       ⋯
 14924 │ P         1357340        1    400     16  1015675     57     85
 14925 │ P         1357340        2    400     16  1015675     69     79
 14926 │ P         1357874        1    900     16  1015675     32     84
 14927 │ P         1357874        2    900     16  1015675     28     74       ⋯
 14928 │ P         1357880        1    500     16  1015675    121     22
 14929 │ P         1357880        2    500     16  1015675    112     22
 14930 │ P         1358490        1    700     16  1015675    353     28
 14931 │ P         1358490        2    700     16  1015675    386     23       ⋯
                                              231 columns and 14910 rows omitted

Since this quiz question is asking us to use a library which allows SQL quiries on DataFrames, we have to import it here

In [216]:
Pkg.add("SQLdf") ; using SQLdf

   Resolving package versions...
   Installed Calculus ──────────────── v0.5.1
   Installed Rmath ─────────────────── v0.7.1
   Installed DualNumbers ───────────── v0.6.8
   Installed WinReg ────────────────── v1.0.0
   Installed SQLite_jll ────────────── v3.45.0+0
   Installed HypergeometricFunctions ─ v0.3.23
   Installed StatsFuns ─────────────── v1.3.0
   Installed Rmath_jll ─────────────── v0.4.0+0
   Installed NaNMath ───────────────── v1.0.2
   Installed SQLite ────────────────── v1.6.0
   Installed StatsAPI ──────────────── v1.7.0
   Installed ShiftedArrays ─────────── v2.0.0
   Installed CategoricalArrays ─────── v0.10.8
   Installed StatsModels ───────────── v0.7.3
   Installed RCall ─────────────────── v0.13.18
   Installed StatsBase ─────────────── v0.34.2
   Installed SQLdf ─────────────────── v0.2.0
    Updating `~/.julia/environments/v1.10/Project.toml`
  [1d9d1391] + SQLdf v0.2.0
    Updating `~/.julia/environments/v1.10/Manifest.toml`
  [49dc2e85] + Calculus v0.5.1
  [

Perform the query using the SQLdf macro - looks pretty damn simply, albeit a bit slow 

In [218]:
@sqldf "select pwgtp1 from quiz2_dataframe where AGEP < 50"

10093×1 DataFrame
   Row │ pwgtp1 
       │ Int64  
───────┼────────
     1 │     87
     2 │     88
     3 │     94
     4 │     91
     5 │    539
     6 │    192
     7 │    153
     8 │    232
     9 │    205
    10 │    226
    11 │    225
   ⋮   │   ⋮
 10084 │     11
 10085 │     23
 10086 │     20
 10087 │     24
 10088 │    160
 10089 │    104
 10090 │     39
 10091 │     28
 10092 │    397
 10093 │    481
10072 rows omitted

## 3. 
Using the same data frame you created in the previous problem, what is the equivalent function to unique(acs$AGEP)

In [220]:
@sqldf "select distinct AGEP from quiz2_dataframe"

91×1 DataFrame
 Row │ AGEP  
     │ Int64 
─────┼───────
   1 │    43
   2 │    42
   3 │    16
   4 │    14
   5 │    29
   6 │    40
   7 │    15
   8 │    28
   9 │    30
  10 │     4
  11 │     1
  ⋮  │   ⋮
  82 │    93
  83 │     8
  84 │    36
  85 │    41
  86 │    44
  87 │    84
  88 │    72
  89 │    73
  90 │    85
  91 │    89
70 rows omitted

## 4. 
How many characters are in the 10th, 20th, 30th and 100th lines of HTML from this page:

http://biostat.jhsph.edu/~jleek/contact.html


In [261]:
url = "http://biostat.jhsph.edu/~jleek/contact.html"

"http://biostat.jhsph.edu/~jleek/contact.html"

In [262]:
download(url, "contact.html")

"contact.html"

In [248]:
#samp = HTTP.get(url, cookies=true);
samp = HTTP.open(

#data = String(samp.body)
#samp
#data

LoadError: MethodError: no method matching open(::String)
You may have intended to import Base.open

Closest candidates are:
  open(::Function, ::Union{String, Symbol}, ::Any, ::Any; kw...)
   @ HTTP ~/.julia/packages/HTTP/bDoga/src/HTTP.jl:579
  open(::Function, ::Union{String, Symbol}, ::Any; ...)
   @ HTTP ~/.julia/packages/HTTP/bDoga/src/HTTP.jl:579
  open(::Function, ::Any; suppress_close_error, verbose, headers, maxframesize, maxfragmentation, kw...)
   @ HTTP ~/.julia/packages/HTTP/bDoga/src/WebSockets.jl:351


In [241]:
counter = 0
for line in eachline(samp) #(open("contact.html"))
        if counter == 10 
            println(line)
        else counter += 1 
    end 
end 

LoadError: MethodError: no method matching eachline(::HTTP.Messages.Response)

Closest candidates are:
  eachline(; ...)
   @ Base io.jl:1063
  eachline(::Base.AbstractCmd; keep)
   @ Base process.jl:338
  eachline(::AbstractString; keep)
   @ Base io.jl:1067
  ...


After some trial and error, I've discovered that it's likely easiest to use the Gumbo.jl package for parsing HTML files https://github.com/JuliaWeb/Gumbo.jl . Iterating through a HTML using the previous methods was inadequate and clunky

In [252]:
Pkg.add("Gumbo") ; using Gumbo

   Resolving package versions...
  No Changes to `~/.julia/environments/v1.10/Project.toml`
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


In [283]:
leekhtml = Gumbo.parsehtml(read("contact.html", String), preserve_whitespace=true)

HTML Document:
<!DOCTYPE html>
HTMLElement{:HTML}:<HTML lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta content="Welcome to Jeff Leek&#39;s Research Group" name="Description"/>
    <meta content="Johns Hopkins University, Bloomberg School of Public Health, Biostatistics, Surrogate Variable Analysis, sva, expression heterogeneity, clinical genomics, reproducibility, high-dimensional data, Top Scoring Pair, tspair " name="Keywords"/>
    <meta content="text/html; charset=iso-8859-1" http-equiv="Content-Type"/>
    <meta content="Global" name="Distribution"/>
    <meta content="index,follow" name="Robots"/>
    <link href="images/favicon.ico" rel="icon" type="image/vnd.microsoft.icon"/>
    <link href="images/PixelGreen.css" rel="stylesheet" type="text/css"/>
    <title>
      jeffrey leek contact
    </title>
    <script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-20898652-1']);
  _gaq.push(['_trackPageview']);

  (fu

In [289]:
counter = 0
for elem in PreOrderDFS(leekhtml.root)
    if counter == 10
            println(attrs(elem))
    return
    else counter += 1 
    end
    println(elem) 
end

<HTML lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml"><head>

<meta content="Welcome to Jeff Leek&#39;s Research Group" name="Description"/>
<meta content="Johns Hopkins University, Bloomberg School of Public Health, Biostatistics, Surrogate Variable Analysis, sva, expression heterogeneity, clinical genomics, reproducibility, high-dimensional data, Top Scoring Pair, tspair " name="Keywords"/>
<meta content="text/html; charset=iso-8859-1" http-equiv="Content-Type"/>
<meta content="Global" name="Distribution"/>
<meta content="index,follow" name="Robots"/>

<link href="images/favicon.ico" rel="icon" type="image/vnd.microsoft.icon"/>
<link href="images/PixelGreen.css" rel="stylesheet" type="text/css"/>



<title>jeffrey leek contact</title>

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-20898652-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async

LoadError: UndefVarError: `attrs` not defined

After a string of unsuccessful attempts, I have opted to simply used some bash magic, to extract a specific line using sed from within julia. 

We can execute commands within julia by enclosing the command in back ticks \`command\`

In [310]:
line_ten = read(`sed -e 10p -n contact.html`, String)

"<meta name=\"Distribution\" content=\"Global\" />\n"

As we can see, the newline character is also present and will therefore be counted - so let's use the **chomp** function to remove this trailing character

In [311]:
chomp(line_ten)

"<meta name=\"Distribution\" content=\"Global\" />"

Now take the length of the line

In [312]:
length(chomp(line_ten))

45

45. Do the remainder of the tasks

In [314]:
line_20 = read(`sed -e 20p -n contact.html`, String)

"<script type=\"text/javascript\">\n"

In [315]:
length(chomp(line_20))

31

In [316]:
line_30 = read(`sed -e 30p -n contact.html`, String)

"  })();\n"

In [317]:
length(chomp(line_30))

7

In [318]:
line_100 = read(`sed -e 100p -n contact.html`, String)

"\t\t\t\t<ul class=\"sidemenu\">\n"

In [319]:
length(chomp(line_100))

25

Answer is 45, 31, 7 and 25 

## 5. 
Read this data set into R and report the sum of the numbers in the fourth of the nine columns.

https://d396qusza40orc.cloudfront.net/getdata%2Fwksst8110.for

Original source of the data: 
http://www.cpc.ncep.noaa.gov/data/indices/wksst8110.for

(Hint this is a fixed width file format)

In [320]:
download("https://d396qusza40orc.cloudfront.net/getdata%2Fwksst8110.for", "question5.for") 

"question5.for"

In [333]:
q = readlines(open("question5.for"))

1258-element Vector{String}:
 " Weekly SST data starts week centered on 3Jan1990"
 ""
 "                Nino1+2      Nino3        Nino34        Nino4"
 " Week          SST SSTA     SST SSTA     SST SSTA     SST SSTA"
 " 03JAN1990     23.4-0.4     25.1-0.3     26.6 0.0     28.6 0.3"
 " 10JAN1990     23.4-0.8     25.2-0.3     26.6 0.1     28.6 0.3"
 " 17JAN1990     24.2-0.3     25.3-0.3     26.5-0.1     28.6 0.3"
 " 24JAN1990     24.4-0.5     25.5-0.4     26.5-0.1     28.4 0.2"
 " 31JAN1990     25.1-0.2     25.8-0.2     26.7 0.1     28.4 0.2"
 " 07FEB1990     25.8 0.2     26.1-0.1     26.8 0.1     28.4 0.3"
 " 14FEB1990     25.9-0.1     26.4 0.0     26.9 0.2     28.5 0.4"
 " 21FEB1990     26.1-0.1     26.7 0.2     27.1 0.3     28.9 0.8"
 " 28FEB1990     26.1-0.2     26.7-0.1     27.2 0.3     29.0 0.8"
 ⋮
 " 23OCT2013     20.5-0.4     24.7-0.2     26.3-0.4     28.7 0.1"
 " 30OCT2013     20.5-0.7     24.8-0.1     26.5-0.2     28.8 0.2"
 " 06NOV2013     21.0-0.4     24.8-0.1     26.6 0.0   

Toying with a function which will take the Vector{String} formatted dataset and input it to a DataFrame - more difficult than it first appears

In [506]:
function vectodf(q)
    #mymatrix = Array(1300, 5){undef}
    #mydata = DataFrame(Matrix{Any}(undef, 1300, 5), ["one", "two", "three", "four", "five"]) 
    mydata = DataFrame(Matrix(zeros(1300, 5)), ["one", "two", "three", "four", "five"]) 
    @show mydata
    dfsize = length(q[5:end])
    col, row = 1,1
    while row <= dfsize
        if col == 6 
            col = 1
            row += 1 
        end 
        for f in collect(q[5:end]) 
            z = split(f) 
            for line in z
            @show String(line)
            push!(mydata[row,col], String(line)) #[row,col]
            col += 1
            end 
        #println(f) 
        end
    end
    mydata 
end 

vectodf (generic function with 2 methods)

In [501]:
mydata = DataFrame(Matrix(zeros(1300, 5)), ["one", "two", "three", "four", "five"]) 


1300×5 DataFrame
  Row │ one      two      three    four     five    
      │ Float64  Float64  Float64  Float64  Float64 
──────┼─────────────────────────────────────────────
    1 │     0.0      0.0      0.0      0.0      0.0
    2 │     0.0      0.0      0.0      0.0      0.0
    3 │     0.0      0.0      0.0      0.0      0.0
    4 │     0.0      0.0      0.0      0.0      0.0
    5 │     0.0      0.0      0.0      0.0      0.0
    6 │     0.0      0.0      0.0      0.0      0.0
    7 │     0.0      0.0      0.0      0.0      0.0
    8 │     0.0      0.0      0.0      0.0      0.0
    9 │     0.0      0.0      0.0      0.0      0.0
   10 │     0.0      0.0      0.0      0.0      0.0
   11 │     0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮
 1291 │     0.0      0.0      0.0      0.0      0.0
 1292 │     0.0      0.0      0.0      0.0      0.0
 1293 │     0.0      0.0      0.0      0.0      0.0
 1294 │     0.0      0.0      0.0      0.0      0.0
 1295 │     0.0      0.0      0.0      0.0      0.0
 1296 │     0.0      0.0      0.0      0.0      0.0
 1297 │     0.0      0.0      0.0      0.0      0.0
 1298 │     0.0      0.0      0.0      0.0      0.0
 1299 │     0.0      0.0      0.0      0.0      0.0
 1300 │     0.0      0.0      0.0      0.0      0.0
                                   1279 rows omitted

In [414]:
q_header = split(q[3])
for f in q_header 
    parse(f, String)
end 
#new_header = "Week"
#push!(new_header, x -> String(x) q_header)
#new_header[1:end] 

LoadError: MethodError: no method matching parse(::SubString{String}, ::Type{String})

Closest candidates are:
  parse(::Type{Union{}}, ::Any...; kwargs...)
   @ Base parse.jl:39


Just trying out how I would create a DataFrame using the correct header 

In [400]:
q5_dataframe = DataFrame(zeros(10, 5), new_header)

10×5 DataFrame
 Row │ Nino1+2  Nino3    Nino34   Nino4    Week    
     │ Float64  Float64  Float64  Float64  Float64 
─────┼─────────────────────────────────────────────
   1 │     0.0      0.0      0.0      0.0      0.0
   2 │     0.0      0.0      0.0      0.0      0.0
   3 │     0.0      0.0      0.0      0.0      0.0
   4 │     0.0      0.0      0.0      0.0      0.0
   5 │     0.0      0.0      0.0      0.0      0.0
   6 │     0.0      0.0      0.0      0.0      0.0
   7 │     0.0      0.0      0.0      0.0      0.0
   8 │     0.0      0.0      0.0      0.0      0.0
   9 │     0.0      0.0      0.0      0.0      0.0
  10 │     0.0      0.0      0.0      0.0      0.0

Restructing the file using bash commands - removing the first 4 rows, then cleaning up the inconsistent columns - some are separated by hyphens and others are spaced apart. This confused me for hours as I assumed that there were only 5 columns, whereby the hyphens stood for measurement error, but in fact there are 9 columns in total. 
```bash
tail -n +5 question5.for | tr - " "  | /
awk '{print $1, $2, $3, $4, $5, $6,
$7, $8, $9}' / 
| tr " " '\t' > question5_cleaned.tsv
``` 

Import the re-shaped data into Julia

In [529]:
cleaned_dataframe = DataFrame(CSV.File("question5_cleaned.tsv", header=false))

1254×9 DataFrame
  Row │ Column1    Column2  Column3  Column4  Column5  Column6  Column7  Colum ⋯
      │ String15   Float64  Float64  Float64  Float64  Float64  Float64  Float ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 03JAN1990     23.4      0.4     25.1      0.3     26.6      0.0     28 ⋯
    2 │ 10JAN1990     23.4      0.8     25.2      0.3     26.6      0.1     28
    3 │ 17JAN1990     24.2      0.3     25.3      0.3     26.5      0.1     28
    4 │ 24JAN1990     24.4      0.5     25.5      0.4     26.5      0.1     28
    5 │ 31JAN1990     25.1      0.2     25.8      0.2     26.7      0.1     28 ⋯
    6 │ 07FEB1990     25.8      0.2     26.1      0.1     26.8      0.1     28
    7 │ 14FEB1990     25.9      0.1     26.4      0.0     26.9      0.2     28
    8 │ 21FEB1990     26.1      0.1     26.7      0.2     27.1      0.3     28
    9 │ 28FEB1990     26.1      0.2     26.7      0.1     27.2      0.3     29 ⋯
   10 │ 07MAR1990     26.7      0.3     26.7      0.2     27.3      0.2     28
   11 │ 14MAR1990     26.1      0.4     26.9      0.2     27.3      0.1     28
  ⋮   │     ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮  ⋱
 1245 │ 06NOV2013     21.0      0.4     24.8      0.1     26.6      0.0     28
 1246 │ 13NOV2013     21.1      0.4     24.7      0.2     26.6      0.0     28 ⋯
 1247 │ 20NOV2013     21.3      0.5     24.8      0.2     26.7      0.1     28
 1248 │ 27NOV2013     21.7      0.4     24.9      0.1     26.7      0.1     28
 1249 │ 04DEC2013     22.0      0.3     24.9      0.2     26.8      0.2     28
 1250 │ 11DEC2013     22.4      0.2     25.1      0.0     26.6      0.0     28 ⋯
 1251 │ 18DEC2013     22.4      0.5     25.2      0.0     26.5      0.1     28
 1252 │ 25DEC2013     22.8      0.4     25.2      0.1     26.4      0.2     28
 1253 │ 01JAN2014     23.7      0.0     25.2      0.2     26.3      0.3     28
 1254 │ 08JAN2014     24.2      0.1     25.1      0.5     26.0      0.5     28 ⋯
                                                 2 columns and 1233 rows omitted

In [530]:
sum(cleaned_dataframe.Column4)

32426.699999999997

The answer is 32426.7 

This took hours longer than I had hoped for, and although the tinkering and trial and error is stimulating and fruitful, knowing where the problem lays is importpant for future encounters - the idea behind reading "fixed-width formats" was new to me, so this unfamiliatiry is partly responsible for the pace of this question. These discussions provide some stimulating notes and hints 
* https://discourse.julialang.org/t/reading-fixed-width-files/30163 
* https://github.com/JuliaData/WeakRefStrings.jl
* https://gist.github.com/vaclavdekanovsky/245bbb0158f3480295b817ba88b77702

It took me aggggggges to notice that the columns for each entry were inconsistently formatted - in future I must peruse and inspect the dataset much more rigorously so avoid these relatively bening problems


We could try to **replace** the "-" delimiter with a space and thus even out the differences - and load this into a DataFrame in a single command - the first few lines of the fill will still cause issues in my guess 
`df = CSV.read(IOBuffer(replace(read("), UInt8('\t') => UInt8(' '))), header=1, delim=" ", ignorerepeated=true, type=Int64, DataFrame)` 